### DISCUSSION TITLE
Hooks, callbacks etc

### DISCUSSION THREAD

**Discussion by jmccabe** [odygrd/quill#183] [Ideas]
I've just been having a brief look at quill, with the focus on whether it could replace spdlog in our code because there are a couple of things we want to be able to do that, to some extent, are more difficult than we'd hoped with spdlog.

The first thing is that, when using a rotating file handler (or "sink" in spdlog terms), we'd like to log a few lines of "header" info in. Until recently, doing that in spdlog was awkward as there was no indication of when the rotate occurred, and the rotating file sink was a final class, with a private "rotate" function, that meant just deriving and handling the rotate slightly differently wasn't possible. Spdlog's latest version adds support for callbacks that help this to be done. Unfortunately, though, quill (unless I'm missing something) seems to suffer from the same restriction; class rotating_file_handler is private and there are no indications that a rotate is about to happen :-(

The other thing would be to take the formatted text and encode/encrypt it before it gets written to a file. While it looks like it may be possible to work around this, somehow (possibly in the same way you need to with spdlog currently), it would be really nice to be able to configure everything 'normally' then, when you're sure it's all working, just be able to pass a callback in where the formatted log message was passed in, and an encoded version of it was returned which was then stored in the file, instead of the original formatted string.

Are either of these the sort of thing that could be considered for quill?

**Comment by odygrd** [comment]
Hello,

I understand the limitation of the RotatingFileHandler. It wouldn't be difficult to change it and let the user handle `rotate` differently. 

I haven't invested too much time into the `Handlers` as it is very easy for the user to just create their own Handler class with the desired behaviour. 

See example [example](https://github.com/odygrd/quill/blob/master/examples/example_custom_handler.cpp)

Having said that a solution to the first issue would be you write you own RotatingFileHandler class handling the rotation as you like. 
As for the encoding/decoding you can write a CustomHandler class like in the above example with e.g. constructor `CustomHandler(std::function<void()> encryption_cb)` 

As I said the RotatingFileHandler can be changed in a future version. If you would like to give it a try right now, is creating your own handler classes enough as a workaround for both issues ?

**Reply by jmccabe** [reply]
Many thanks for your reply. I understand what you're suggesting; it's the same principle as that suggested for spdlog. I guess my main issue is that almost all the functionality I need is there in RotatingFileHandler; the additional functionality I need is relatively small, but would essentially mean creating a copy RotatingFileHandler to  achieve it, rather than just using the decorator pattern (because I can't wrap RotatingFileHandler::_rotate(), as it's private), or the Template Method Pattern, which I can't use as RotatingFileSink is final.

For what it's worth, though, I had a separate issue. We are trying to target two systems; Ubuntu, and another embedded Linux target. Due to certain dependencies we're stuck with gcc 7.3.1 or so on the embedded target and are using 7.5.0 on Ubuntu, which is mostly the same (!). I tried some simple examples on Ubuntu, which worked as expected, but when I did the same thing using quill built with the embedded target's compiler, I didn't have much success. I had 3 handlers registered as defaults, two file handlers with different log levels, and a colour-supporting std out handler. Unfortunately, when I ran it on the target, the files got created, but remained empty, no log entries were shown on stdout, and I got an error raised in line 604 of the bundled fmt's core.h when the program exited. There were a number of warnings during the build of quill, admittedly, but it's nominally the same version of fmt that's bundled with spdlog 1.10.0 which didn't show those warnings when it was built. At the moment I can't really spare much more time to investigate this but, if I do, I'll let you know if there's anything I can do.

Cheerz
John
